### PR TITLE
[c++] Addition of `SOMADataFrame`

### PR DESCRIPTION
**Pull request by nguyenv**
**Issue and/or context:**
https://github.com/single-cell-data/TileDB-SOMA/issues/1257

**Changes:**
* Addition of `SOMADataFrame` which inherits from `SOMAArray`

**Notes for Reviewer:**
N/A

### CODE REVIEW

File: libtiledbsoma/src/soma/soma_array.h [hunk 1]
```diff
@@ -128,6 +139,20 @@ class SOMAArray {
     SOMAArray(SOMAArray&&) = default;
     ~SOMAArray() = default;
 
+    /**
+     * @brief Get URI of the SOMAArray.
+     *
+     * @return std::string URI
+     */
+    std::string uri() const;
```
Comment (teo-tsirpanis) [inline libtiledbsoma/src/soma/soma_array.h:147]:
```suggestion
    const std::string& uri() const;
```

File: libtiledbsoma/src/soma/soma_dataframe.h [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+     * @param platform_config Platform-specific options used to create this
+     * DataFrame
+     * @param column_names A list of column names to use as user-defined index
+     * columns (e.g., ``['cell_type', 'tissue_type']``). All named columns must
+     * exist in the schema, and at least one index column name is required.
+     * @param timestamp If specified, overrides the default timestamp used to
+     * open this object. If unset, uses the timestamp provided by the context.
+     * @return std::shared_ptr<SOMADataFrame> SOMADataFrame
+     */
+    static std::shared_ptr<SOMADataFrame> open(
```
Comment (teo-tsirpanis) [inline libtiledbsoma/src/soma/soma_dataframe.h:76]: For a later time, all these `open` functions could return a `unique_ptr`, which can be moved to a `shared_ptr` if needed by the caller. But it doesn't have to be done now.

**Review by eddelbuettel — APPROVED** [review]
Seems good to go but part of me wonders if we should involve Python and/or R in the testing?  The library has well-established clients.

File: libtiledbsoma/src/soma/soma_dataframe.h [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+     * @param mode TILEDB_READ or TILEDB_WRITE
+     * @param uri URI of the array
+     * @param name name of the array
+     * @param ctx TileDB context
+     * @param column_names Columns to read
+     * @param batch_size Batch size
+     * @param result_order Result order
+     * @param timestamp Timestamp
+     */
+    SOMADataFrame(
```
[the remark this answers is not in the document]
Comment (nguyenv) [reply libtiledbsoma/src/soma/soma_dataframe.h:121]: Why wouldn't we be opening and closing `SOMADataFrame` in the external code?

[the remark this answers is not in the document]
Comment (teo-tsirpanis) [reply libtiledbsoma/src/soma/soma_dataframe.h:121]: It's not compliant with the [C.41](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c41-a-constructor-should-create-a-fully-initialized-object) guideline. Once you create a dataframe you cannot immediately use it; you first have to `open` it, which is not immediately apparent.

[the remark this answers is not in the document]
Comment (nguyenv) [reply libtiledbsoma/src/soma/soma_dataframe.h:118]: OIC. I just update `create` to return a dataframe opened in read mode and updated the documentation to reflect that.

**Pull request merged**